### DISCUSSION TITLE
Cache issue React

### DISCUSSION THREAD

**Discussion by BenchBadr** [vercel/vercel#11735] [Help]
### Question

Hello, I have a React project hosted on Vercel and whenever I push to prod I have the same issue : white screen until I clear the cache.
After clearing cache it works as expected but I do not expect all my users to do this whenever I do some edits on the website.
I tried to edit `vercel.json` to disable caching but it doesn't work quite well.
Can you please enlighten me about how to fix this issue ?

```json
{
  "builds": [
    {
      "src": "app.py",
      "use": "@vercel/python"
    }
  ],
  "rewrites": [
    {
      "source": "/(.*)",
      "destination": "app.py"
    }
  ],
  "headers": [
    {
      "source": "/(.*)",
      "headers": [
        {
          "key": "Cache-Control",
          "value": "no-store"
        }
      ]
    }
  ]
}
```

<img width="438" alt="image" src="https://github.com/vercel/vercel/assets/78297845/1dd285a2-5edf-4446-85f8-b5702c410edc">

Thanks in advance!

**Comment by H-Hp** [comment]
The issue you're facing is related to browser caching, which can sometimes cause outdated content to be displayed even after deploying new changes. This typically happens when the browser caches static assets (HTML, CSS, JavaScript files) and serves the cached version instead of fetching the latest version from the server.

To address this issue in a React project hosted on Vercel, you can try the following approaches:

1. **Leverage Vercel's Cache-Control Headers**:
   Vercel provides a way to set Cache-Control headers for your deployments. These headers instruct the browser and intermediate caches how to handle caching for your assets.

   In your `vercel.json` file, you can set the `Cache-Control` header to `no-cache` or `no-store` for all routes or specific routes. This will prevent the browser from caching the assets and force it to fetch the latest version on each request.

   Here's an example of how you can set the `Cache-Control` header for all routes in your `vercel.json` file:

   ```json
   {
     "headers": [
       {
         "source": "/(.*)",
         "headers": [
           {
             "key": "Cache-Control",
             "value": "no-cache"
           }
         ]
       }
     ]
   }
   ```

   Replace `"value": "no-cache"` with `"value": "no-store"` if you want to completely disable caching.

2. **Implement Cache-Busting Techniques**:
   Another approach is to implement cache-busting techniques in your React application. These techniques ensure that the browser fetches the latest version of your assets by appending a unique identifier (e.g., a hash or version number) to the asset URLs.

   One common way to achieve this is by using a tool like `react-script` or `webpack` to generate a unique hash for each build and append it to the filenames of your static assets. This way, whenever you deploy a new version, the asset URLs will be different, forcing the browser to fetch the new assets.

   Alternatively, you can use a library like `react-hot-loader` or `react-refresh` to enable hot module replacement (HMR) during development. With HMR, the browser will automatically update the changed modules without requiring a full page refresh, which can help mitigate caching issues.

3. **Implement Service Workers (PWA)**:
   If your React application is a Progressive Web App (PWA), you can leverage service workers to control caching behavior. Service workers can intercept network requests and serve cached responses or fetch fresh content from the server based on your caching strategy.

   With service workers, you can implement various caching strategies, such as cache-first (serve cached content first, and fetch fresh content in the background), network-first (fetch fresh content first, and cache for later use), or a hybrid approach that combines both strategies.

4. **Use Vercel's Incremental Static Regeneration (ISR)**:
   If your React application uses static site generation (SSG) with tools like Next.js, you can take advantage of Vercel's Incremental Static Regeneration (ISR) feature. ISR allows you to update static pages after deploying without requiring a full rebuild, which can help mitigate caching issues.

   With ISR, you can define a revalidation period for your static pages, and Vercel will automatically rebuild and update the pages when the revalidation period expires. This ensures that users always receive the latest content without having to clear their cache manually.

By implementing one or more of these approaches, you should be able to mitigate the white screen issue caused by browser caching and ensure that your users receive the latest version of your React application without having to manually clear their cache.
